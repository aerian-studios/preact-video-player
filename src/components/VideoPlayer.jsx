import { h, Component } from 'preact'
import PropTypes from 'proptypes'
import videojs from 'video.js'

/* eslint-disable import/no-webpack-loader-syntax,import/no-unresolved */
import '!style-loader!css-loader!video.js/dist/video-js.css'

export default class VideoPlayer extends Component {
  constructor(...props) {
    super(...props)
    this.handleReady = this.handleReady.bind(this)
    this.handleRef = this.handleRef.bind(this)
  }

  componentDidMount() {
    this.player = videojs(this.videoNode, this.props.options, this.handleReady)
  }

  componentWillUnmount() {
    if (this.player) {
      this.player.dispose()
    }
  }

  handleReady() {
    console.log('ready', this.player)
  }

  handleRef(node) {
    this.videoNode = node
  }

  // wrap the player in a div with a `data-vjs-player` attribute
  // so videojs won't create additional wrapper in the DOM
  // see https://github.com/videojs/video.js/pull/3856
  /* eslint-disable jsx-a11y/media-has-caption */
  render() {
    return (
      <div>
        <div data-vjs-player>
          <video ref={this.handleRef} className="video-js" />
        </div>
      </div>
    )
  }
}

VideoPlayer.propTypes = {
  options: PropTypes.shape({
    autoplay: PropTypes.bool,
    controls: PropTypes.bool,
    sources: PropTypes.array({
      src: PropTypes.string,
      type: PropTypes.string
    })
  }).isRequired
}
