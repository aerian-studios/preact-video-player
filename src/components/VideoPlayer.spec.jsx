import { h } from 'preact'
import { shallow } from 'preact-render-spy'
import videojs from 'video.js'
import VideoPlayer from './VideoPlayer'

jest.mock('video.js')

videojs.dispose = jest.fn()

describe('<VideoPlayer />', () => {
  it('should render', () => {
    const context = shallow(<VideoPlayer />)
    expect(context.toString()).toMatchSnapshot()
  })

  it('should initialize video.js on mount', () => {
    shallow(<VideoPlayer />)
    expect(videojs).toBeCalled()
  })

  it('should dispose video.js instance on unmount', () => {
    const context = shallow(<VideoPlayer />)
    const disposeSpy = jest.fn()
    context.component().player = { dispose: disposeSpy }
    context.component().componentWillUnmount()
    expect(disposeSpy).toHaveBeenCalled()
  })
})
