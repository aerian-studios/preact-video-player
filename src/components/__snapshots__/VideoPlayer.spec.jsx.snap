// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoPlayer /> should render 1`] = `
"preact-render-spy (1 nodes)
-------
<div>
  <div data-vjs-player={true}>
    <video class=\\"video-js\\"></video>
  </div>
</div>
"
`;
